File: Hash/benchmark/go/TestHash.go
package main

import (
	"crypto/md5"
	"crypto/sha1"
	"crypto/sha256"
	"crypto/sha512"
	"golang.org/x/crypto/ripemd160"
	"golang.org/x/crypto/sha3"

	"bytes"
	"encoding/binary"
	"fmt"
	"hash"
	"io/ioutil"
	"math/rand"
	"time"
)

func main() {
	//生成多个随机测试数据，分别是1字节、1MB、30MB、90MB
	//ioutil.WriteFile("1B.txt", []byte("1"), 0644)
	//generateRanddata("1KB.txt", 25)
	//generateRanddata("1MB.txt", 720)
	//generateRanddata("30MB.txt", 645*6)
	//generateRanddata("90MB.txt", 671*10)

	//开始测试
	//files := []string{"1B.txt","1KB.txt","1MB.txt","30MB.txt"}
	files := []string{"1B.txt","1KB.txt"}

	for _,v :=range files{
		fmt.Println("\n",v)
		dat,_ := ioutil.ReadFile(v)//读取相应文件中的信息
		plain := []byte(dat)
		fmt.Println(len(plain))

		//fmt.Print("MD5: ")
		md := md5.New()
		testMD(md, plain)

		//fmt.Print("SHA-1: ")
		md = sha1.New()
		testMD(md, plain)

		//fmt.Print("SHA224: ")
		md = sha256.New224()
		testMD(md, plain)

		//fmt.Print("SHA256: ")
		md = sha256.New()
		testMD(md, plain)

		//fmt.Print("SHA384: ")
		md = sha512.New384()
		testMD(md, plain)

		//fmt.Print("SHA512: ")
		md = sha512.New()
		testMD(md, plain)

		//fmt.Print("SHA512/224: ")
		md = sha512.New512_224()
		testMD(md, plain)

		//fmt.Print("SHA512/256: ")
		md = sha512.New512_256()
		testMD(md, plain)

		//fmt.Print("RIPEMD160: ")
		ripe := ripemd160.New()
		testMD(ripe, plain)

		//fmt.Print("SHA3-224: ")
		md = sha3.New224()
		testMD(md, plain)

		//fmt.Print("SHA3-256: ")
		md = sha3.New256()
		testMD(md, plain)

		//fmt.Print("SHA3-384: ")
		md = sha3.New384()
		testMD(md, plain)

		//fmt.Print("SHA3-512: ")
		md = sha3.New512()
		testMD(md, plain)

		//fmt.Print("SHAKE128: ")
		//testShake128(plain)
		//
		//fmt.Print("SHAKE256: ")
		//testShake256(plain)

	}
}

func testMD(md hash.Hash, msg []byte) time.Duration {
	//fmt.Println("输出大小", md.Size())

	t1 := time.Now()

	for i := 0; i < 10000; i++ {
		md.Reset()
		md.Write(msg)
		md.Sum(nil)
		//fmt.Printf("%x", md.Sum(nil))
		//fmt.Println(" ")
	}
	t2 := time.Now()
	fmt.Println(t2.Sub(t1))
	return t2.Sub(t1)
}



func check(e error){
	if e!=nil{
		panic(e)
	}
}

func generateRanddata(filename string, round int){
	rand.Seed(time.Now().Unix())
	buf := bytes.NewBuffer([]byte{})
	var data []byte

	for index:=0; index<round; index++{
		x := rand.Int()
		binary.Write(buf, binary.BigEndian, int32(x))
		data = append(data, buf.Bytes()...)
	}

	err:= ioutil.WriteFile(filename, data, 0644)
	check(err)
	//fmt.Println(len(data))
}

File: Hash/benchmark/js/testHash.js
// 本地需安装NodeJS环境，该文档是在8.11.3版本下测试的，高版本应该也可以使用
// 依赖包 npm install js-sha3
// node testhash.js

const crypto = require('crypto');
const fs = require('fs');
const sha3js = require('js-sha3');

var array = ["1B.txt","1KB.txt","1MB.txt","30MB.txt"];
//var array = ["1MB.txt"];
const N = 10000
array.forEach(function(v,i,a){
        console.log(v);
        var data = fs.readFileSync(v);

        //md5 sha1 sha256 sha512 ripemd160
        testEachHash("md5", data)
        testEachHash("sha1", data)
        testEachHash("sha224", data)
        testEachHash("sha256", data)
        testEachHash("SHA384", data)
        testEachHash("sha512", data)
        testEachHash("ripemd160", data)

        testsha3224(data)
        testsha3256(data)
        testsha3384(data)
        testsha3512(data)
});

function testsha3224(data){
    t1 = new Date().getTime();
    for( var i = 0; i<N; i++) {
        var hash = sha3js.sha3_224.create();
        hash.update(data);
        //console.log(hash.hex());
        hash.digest('hex');
    }
    t2 = new Date().getTime()
    interval = (t2 - t1)/1000;
    console.log("sha3-224: ", interval, "s")
}

function testsha3256(data){
    t1 = new Date().getTime();
    for( var i = 0; i<N; i++) {
        var hash = sha3js.sha3_256.create();
        hash.update(data);
        //console.log(hash.hex());
        hash.digest('hex');
    }
    t2 = new Date().getTime()
    interval = (t2 - t1)/1000;
    console.log("sha3-256: ", interval, "s")
}

function testsha3384(data){
    t1 = new Date().getTime();
    for( var i = 0; i<N; i++) {
        var hash = sha3js.sha3_384.create();
        hash.update(data);
        //console.log(hash.hex());
        hash.digest('hex');
    }
    t2 = new Date().getTime()
    interval = (t2 - t1)/1000;
    console.log("sha3-384: ", interval, "s")
}

function testsha3512(data){
    t1 = new Date().getTime();
    for( var i = 0; i<N; i++) {
        var hash = sha3js.sha3_512.create();
        hash.update(data);
        //console.log(hash.hex());
        hash.digest('hex');
    }
    t2 = new Date().getTime()
    interval = (t2 - t1)/1000;
    console.log("sha3-512: ", interval, "s")
}

function testEachHash(hashname, data) {
    t1 = new Date().getTime();
    for( var i = 0; i<N; i++) {
        const hash = crypto.createHash(hashname);
        hash.update(data)
        //console.log(hash.digest('hex'));
        hash.digest('hex');
    }
    t2 = new Date().getTime()
    interval = (t2 - t1)/1000;
    console.log(hashname, ": ", interval, "s")
}


